### PR TITLE
Fix remaining method ambiguities

### LINKED ISSUE #2261: Enable Aqua.test_ambiguities
In #2257 I disabled method ambiguity checking, but @devmotion [suggested](https://github.com/TuringLang/Turing.jl/pull/2257#discussion_r1630240959) enabling them with
```suggestion
@testset "Aqua" begin
    # Test ambiguities separately without Base and Core
    # Ref: https://github.com/JuliaTesting/Aqua.jl/issues/77
    Aqua.test_all(Turing; ambiguities = false)
    Aqua.test_ambiguities(Turing)
end
```

We should do that and see if we can fix the 6 ambiguities that it warns about. I'm happy to do this, but off next week, so will take a moment.

### PR DESCRIPTION
**Pull request by penelopeysm**
This PR:

1. Enables Aqua method ambiguity tests (as per #2290)
2. Creates an `AbstractTransition` abstract type which all the `Transitions` in Turing subtype.
3. Modifies the type signature of `bundle_samples` (in `src/mcmc/Inference.jl`) to take a `Vector{<:Union{AbstractTransition,AbstractVarInfo}}` as the first argument. The `AbstractVarInfo` case occurs when sampling with `Prior()`, so the type signature of this argument somewhat mirrors that of the `sampler` argument in the same function (which is `Union{Sampler{<:InferenceAlgorithm},SampleFromPrior}`).
4. Modifies the type signature of `Base.get` (in `src/optimisation/Optimisation.jl`) to take an `AbstractVector{Symbol}` as the second argument.

(4) is technically a breaking change because it is a more restrictive type signature, hence I've also bumped the version number to 0.34.0.

Closes #2261; extends #2290.

### CODE REVIEW

**Comment by penelopeysm** [comment]
All tests on Julia 1.7 are failing, I suspect it's because ADTypes resolves to 0.2.7 (as opposed to 1.7.1 on Julia 1.10), and that breaks when used with the new v.1.9.1 of LogDensityProblemsAD.

Not sure why this is the case — I'd have thought that the updated compat entry in https://github.com/tpapp/LogDensityProblemsAD.jl/pull/34 would have prevented this from happening.

**Comment by mhauru** [comment]
@sunxd3, @yebai, any thoughts on the above?

**Comment by devmotion** [comment]
My gut feeling is that right now an `AbstractTransition` has only limited use since all these transitions have different fields and even different field names for the samples (e.g., different in Turing compared with AdvancedMH), and hence if you dispatch on it as in the `bundle_samples` functions you must rely on some (Turing-internal) functions for extracting log evidence etc.

I think it could be useful to add an interface for accessing the most important properties of a transition to AbstractMCMC + a default `Transition` implementation (it seems, in Turing and AdvancedMH it's mainly samples, log density + some statistics) that downstream packages could use instead of defining their own types. Possibly, in most cases such a default `Transition` object could be sufficient? If not, I think AbstractMCMC would also be a natural location for an abstract supertype such as `AbstractTransition`.

File: src/optimisation/Optimisation.jl [hunk 1]
```diff
@@ -277,13 +277,13 @@ StatsBase.loglikelihood(m::ModeResult) = m.lp
 
 """
     Base.get(m::ModeResult, var_symbol::Symbol)
-    Base.get(m::ModeResult, var_symbols)
+    Base.get(m::ModeResult, var_symbols::AbstractVector{Symbol})
```
[the remark this answers is not in the document]
Comment (penelopeysm) [reply src/optimisation/Optimisation.jl:280]: It could be! Would you prefer this type to be expanded to `Union{AbstractVector{Symbol},Tuple{Symbol}}`?

**Comment by penelopeysm** [comment]
I agree that the fields of the different `Transition`s don't define a common interface (yet) — the only usefulness of defining an abstract supertype would be to fix the method ambiguity and maybe a small improvement in code readability.

I'm in favour of having an overarching `AbstractTransition` type in AbstractMCMC! I think, though, that will run a bigger risk of breaking some of the code examples that @mhauru brought up, and so it'd need to be done very carefully as a larger piece of work.

@mhauru: Regarding your point about other libraries: Right now some libraries in the Turing ecosystem have their own kind of `Transition`, e.g. AdvancedHMC and AdvancedMH, and some of them also define their own version of `bundle_samples` that dispatch on their own `Transition` type. I think the way this works out is that they will either call their own version of `bundle_samples` _or_ (if it's not defined) the default AbstractMCMC implementation, but not an implementation within Turing core.

I'm less certain about this, but I think it's also not possible to call Turing's `bundle_samples` with a `Transition` from a different library as Turing's `bundle_samples` is restricted to a certain set of samplers that are defined within Turing itself (and the samplers will define the `Transition` type).

**Pull request merged**